src/briefcase/commands/run.py

```diff
@@ -1,23 +1,273 @@
+        r") in \d+\.\d+s ={1,}$)"
+    )
+
+    DEFAULT_FAILURE_REGEX = (
+        # Unittest
+        # The ?P<a...> groups are to support Android logging differences.
+        # when https://github.com/chaquo/chaquopy/issues/746 is resolved,
+        # those groups can be deleted.
+        r"(^-{65,}\n(?P<a1> \n)?Ran \d+ tests in \d+\.\d{3}s"
+        r"\n(?P<a2> )?\n(?P<a3> \n)?FAILED( \(.*\))?$)"
+        # Pytest
+        r"|"
+        r"(^={1,} ("
+        r"((\d+ failed)((, )?\d+ (passed|skipped|deselected|xfailed|xpassed|warnings?|errors?))*)"
+        r"|(((, )?\d+ (failed|passed|skipped|deselected|xfailed|xpassed|warnings?|errors?))*((, )?\d+ errors?))"
+        r") in \d+.\d+s ={1,}$)"
+    )
+
+    def __init__(
+        self,
+        log_popen,
+        clean_filter,
+        clean_output,
+        success_filter,
+        failure_filter,
+    ):
+        """Create a filter for a log stream.
+
+        :param log_popen: The Popen object for the stream producing the logs.
+        :param clean_filter: A function that will filter a line of logs,
+            returning a "clean" line without any log system preamble.
+        :param clean_output: Should the output displayed to the user be the
+            "clean" output? (Default: True).
+        :param success_filter: A function that will operate on a string
+            containing the last 10 lines of "clean" (i.e., preamble filtered)
+            logs, returning True if a "success" condition has been detected. If
+            the success filter returns True, the log process will be terminated.
+        :param failure_filter: A function that will operate on a string
+            containing the last 10 lines of "clean" (i.e., preamble filtered)
+            logs, returning True if a "failure" condition has been detected. If
+            the failure filter returns True, the log process will be terminated.
+        """
+        self.log_popen = log_popen
+        self.success = None
+        self.clean_filter = clean_filter
+        self.clean_output = clean_output
+
+        self.recent_history = []
+        self.success_filter = success_filter
+        self.failure_filter = failure_filter
+
+    def __call__(self, line):
+        """Filter a single line of a log.
+
+        :param line: A single line of raw system log content, including the newline.
+        """
+        # Compute the clean line
+        if self.clean_filter:
+            filtered = self.clean_filter(line)
+
+            # If the clean filter says the line can be dumped, return None
+            if filtered is None:
+                return None
+
+            # If there's a cleaned line, we can determine if it should be included in analysis
+            clean_line, included = filtered
+        else:
+            # If we don't perform cleaning, we assume all content is potentially
+            # Python, and should be included in analysis
+            clean_line = line
+            display_line = line
```

Comment:
Setting `display_line = line` here is functionally a no-op due to the following `if/else`.